File: license/license.go
/* {{{ Copyright (c) Paul R. Tagliamonte <paultag@opensource.org>, 2015-2016
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE. }}} */

package license

import (
	"encoding/json"
	"fmt"
	"os"
)

//
type Identifier struct {
	Identifier string `json:"identifier"`
	Scheme     string `json:"scheme"`
}

//
type Link struct {
	Note *string `json:"note"`
	URL  string  `json:"url"`
}

//
type OtherName struct {
	Name string  `json:"name"`
	Note *string `json:"note"`
}

//
type Text struct {
	MediaType string `json:"media_type"`
	Title     string `json:"title"`
	URL       string `json:"url"`
}

//
type License struct {
	Id           string       `json:"id"`
	Identifiers  []Identifier `json:"identifiers"`
	Links        []Link       `json:"links"`
	Name         string       `json:"name"`
	OtherNames   []OtherName  `json:"other_names"`
	SupersededBy *string      `json:"superseded_by"`
	Keywords     []string     `json:"keywords"`
	Texts        []Text       `json:"text"`
}

//
type Licenses []License

//
func (licenses Licenses) GetIds() []string {
	identifiers := []string{}
	for _, license := range licenses {
		identifiers = append(identifiers, license.Id)
	}
	return identifiers
}

//
func (licenses Licenses) GetIdMap() map[string]License {
	ret := map[string]License{}
	for _, license := range licenses {
		ret[license.Id] = license
		for _, identifier := range license.Identifiers {
			ret[fmt.Sprintf("%s/%s", identifier.Scheme, identifier.Identifier)] = license
		}
	}
	return ret
}

//
func (licenses Licenses) GetTagMap() map[string][]License {
	ret := map[string][]License{}
	for _, license := range licenses {
		for _, tag := range license.Keywords {
			ret[tag] = append(ret[tag], license)
		}
	}
	return ret
}

//
func LoadLicensesFiles(path string) (Licenses, error) {
	ret := Licenses{}
	fh, err := os.Open(path)
	if err != nil {
		return Licenses{}, err
	}
	if err := json.NewDecoder(fh).Decode(&ret); err != nil {
		return Licenses{}, err
	}
	return ret, nil
}

// vim: foldmethod=marker

File: license/License.hs
{- Haskell code for accessing OSI license API

Copyright © 2016 Clint Adams

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE
-}

{-# LANGUAGE TemplateHaskell #-}

module License (
    OSIIdentifier(..)
  , OSILink(..)
  , OSIOtherName(..)
  , OSIText(..)
  , OSILicense(..)
  , allLicenses
  , licensesMatchingKeyword
  , licenseById
  , licenseBySchemeAndIdentifier
) where

import Control.Monad.Trans.Except (ExceptT(..))
import Data.Aeson (eitherDecode)
import Data.Aeson.TH (defaultOptions, deriveJSON, Options(..))
import Data.Char (toLower)
import Data.Text (Text)
import Network.HTTP.Client (httpLbs, newManager, parseUrl, responseBody)
import Network.HTTP.Client.TLS (tlsManagerSettings)

data OSIIdentifier = OSIIdentifier {
    oiIdentifier :: Text
  , oiScheme :: Text
} deriving (Eq, Read, Show)
$(deriveJSON defaultOptions{fieldLabelModifier = (map toLower . drop 2), constructorTagModifier = map toLower} ''OSIIdentifier)

data OSILink = OSILink {
    olNote :: Text
  , olUrl :: Text
} deriving (Eq, Read, Show)
$(deriveJSON defaultOptions{fieldLabelModifier = (map toLower . drop 2), constructorTagModifier = map toLower} ''OSILink)

data OSIOtherName = OSIOtherName {
    oonName :: Text
  , oonNote :: Maybe Text
} deriving (Eq, Read, Show)
$(deriveJSON defaultOptions{fieldLabelModifier = (map toLower . drop 3), constructorTagModifier = map toLower} ''OSIOtherName)

data OSIText = OSIText {
    otMedia_type :: Text
  , otTitle :: Text
  , otURL :: Text
} deriving (Eq, Read, Show)
$(deriveJSON defaultOptions{fieldLabelModifier = (map toLower . drop 2), constructorTagModifier = map toLower} ''OSIText)

data OSILicense = OSILicense {
    olId :: Text
  , olName :: Text
  , olSuperseded_by :: Maybe Text
  , olKeywords :: [Text]
  , olIdentifiers :: Maybe [OSIIdentifier]  -- BUG in API response?
  , olLinks :: [OSILink]
  , olOther_names :: Maybe [OSIOtherName]   -- BUG in API response?
  , olText :: [OSIText]
} deriving (Eq, Read, Show)
$(deriveJSON defaultOptions{fieldLabelModifier = (map toLower . drop 2), constructorTagModifier = map toLower} ''OSILicense)

getLicenses :: String -> ExceptT String IO [OSILicense]
getLicenses k = ExceptT $ do
  manager <- newManager tlsManagerSettings
  request <- parseUrl $ "https://api.opensource.org/licenses/" ++ k
  response <- httpLbs request manager

  return . eitherDecode . responseBody $ response

allLicenses :: ExceptT String IO [OSILicense]
allLicenses = getLicenses ""

licensesMatchingKeyword :: String -> ExceptT String IO [OSILicense]
licensesMatchingKeyword = getLicenses

getLicense :: String -> ExceptT String IO OSILicense
getLicense k = ExceptT $ do
  manager <- newManager tlsManagerSettings
  request <- parseUrl $ "https://api.opensource.org/license/" ++ k
  response <- httpLbs request manager

  return . eitherDecode . responseBody $ response

licenseById :: String -> ExceptT String IO OSILicense
licenseById = getLicense

licenseBySchemeAndIdentifier :: String -> String -> ExceptT String IO OSILicense
licenseBySchemeAndIdentifier s i = getLicense (concat [s, "/", i])
